src/java.base/share/classes/jdk/internal/misc/CDS.java

```diff
@@ -59,4 +78,125 @@
      * Check if sharing is enabled via the UseSharedSpaces flag.
      */
     public static native boolean isSharingEnabled();
+
+    /**
+     * log lambda form invoker holder, name and method type
+     */
+    public static void traceLambdaFormInvoker(String prefix, String holder, String name, String type) {
+        if (isDumpingClassList) {
+            logLambdaFormInvoker(prefix + " " + holder + " " + name + " " + type);
+        }
+    }
+
+    /**
+      * log species
+      */
+    public static void traceSpeciesType(String prefix, String cn) {
+        if (isDumpingClassList) {
+            logLambdaFormInvoker(prefix + " " + cn);
+        }
+    }
+
+    static final String DIRECT_HOLDER_CLASS_NAME  = "java.lang.invoke.DirectMethodHandle$Holder";
+    static final String DELEGATING_HOLDER_CLASS_NAME = "java.lang.invoke.DelegatingMethodHandle$Holder";
+    static final String BASIC_FORMS_HOLDER_CLASS_NAME = "java.lang.invoke.LambdaForm$Holder";
+    static final String INVOKERS_HOLDER_CLASS_NAME = "java.lang.invoke.Invokers$Holder";
+
+    private static boolean isValidHolderName(String name) {
+        return name.equals(DIRECT_HOLDER_CLASS_NAME)      ||
+               name.equals(DELEGATING_HOLDER_CLASS_NAME)  ||
+               name.equals(BASIC_FORMS_HOLDER_CLASS_NAME) ||
+               name.equals(INVOKERS_HOLDER_CLASS_NAME);
+    }
+
+    private static boolean isBasicTypeChar(char c) {
+         return "LIJFDV".indexOf(c) >= 0;
+    }
+
+    private static boolean isValidMethodType(String type) {
+        String[] typeParts = type.split("_");
+        // check return type (second part)
+        if (typeParts.length != 2 || typeParts[1].length() != 1
+                || !isBasicTypeChar(typeParts[1].charAt(0))) {
+            return false;
+        }
+        // first part
+        if (!isBasicTypeChar(typeParts[0].charAt(0))) {
+            return false;
+        }
+        for (int i = 1; i < typeParts[0].length(); i++) {
+            char c = typeParts[0].charAt(i);
+            if (!isBasicTypeChar(c)) {
+                if (!(c >= '0' && c <= '9')) {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    // return null for invalid input
+    private static Stream<String>  validateInputLines(String[] lines) {
+        ArrayList<String> list = new ArrayList<String>(lines.length);
+        for (String s: lines) {
+            String line = s.trim();
+            if (!line.startsWith("[LF_RESOLVE]") && !line.startsWith("[SPECIES_RESOLVE]")) {
+                System.out.println("Wrong prefix: " + line);
```

Comment:
Should this throw an exception instead?

Reply:
This part is for check the format only, throw exceptions will lead more objects generated which should not be archived in shared heap. Since this is only called from VM, so decide not to throw exception here.

Reply:
The exception object will not be automatically added to the shared heap, so it's OK to throw exceptions here.